services/matching-service/start.ts

```diff
@@ -0,0 +1,120 @@
+        }
+        
+        //Check valid Category
+        if (category == undefined) {
+            socket.emit("error", "Invalid Category!");
+            return;
+        }
+        
+        // Check for duplicate ID and push to Queue
+        if (checkIdExists(msg.id)) {
+            socket.emit("error", "You are already in the queue!");
+            return;
+        } 
+
+        // Find match if exists, else create new queue and wait.
+        const this_match = {
+                id: msg.id,
+                difficulty: difficulty,
+                category: category,
+                sockAddr: socket.id
+            }
+        const queueName = difficulty + category; // e.g. easyrecursion
+        let this_queue = queues.get(queueName);
+        if (this_queue) {
+            // Match found, remove queue from queues
+            this_queue.push(this_match)
+            console.log("WE FOUND A MATCH!");
+            queues.delete(queueName);
+            
+            // Remove id from ids.
+            const peerID = this_queue[0].id;
+            ids = ids.filter(id => id != msg.id && id != peerID);
+            
+            // Emit on this user's socket
+            socket.emit("success", `You have been paired with User ${peerID}.`);
+
+            // Emit on paired user's socket
+            const peerSockAddr = this_queue[0].sockAddr;
+            const peerSocket = io.sockets.sockets.get(peerSockAddr);
+            if (peerSocket) {
+                peerSocket.emit("success", `You have been paired with User ${this_queue[1].id}`);
+                peerSocket.disconnect();
+            } else {
+                console.log("Cannot find peer socket!");
+            }        
+            socket.disconnect();
+        } else {
+            // No match, so we push queue to queues.
+            this_queue = [this_match];
+            queues.set(queueName, this_queue);
+
+            // Push this user's id into ids.
+            ids.push(msg.id);
+
+            // Wait
+            const DELAY = 5000;
+            setTimeout(() => {
+                this_queue = queues.get(queueName);
+                if (ids.includes(msg.id) && this_queue) {
+                    queues.delete(queueName);
+                    ids = ids.filter(id => id != msg.id);
+                    console.log("FAILURE!");
+                    socket.emit("error", "Sorry, we could not find you a match!");
+                    socket.disconnect();
+                }
+            }, DELAY);
+        }
+        socket.emit("reply", "You are in the queue!");
+    })
+
+
+
+
+
+    socket.on("disconnect", () => {
+        console.log("User has disconnected!");
+    })
+});
+
+httpServer.listen(port, () => {
+    console.log(`Matching service running on ${port}`)
+});
+
+
+export interface Match {
+  id: number,
+  difficulty: string,
+  category: string,
+  sockAddr: string
+}
+
+const queues: Map<string, Array<Match>> = new Map();
+
+export let ids: Array<number> = [];
```

Comment:
Can use a `Set` instead of `Array` for `ids`. We can use `Set.delete(id)` instead of filtering the array

Reply:
Admittedly, Array of ids was a bad idea. I stopped tracking the ```ids```. If we need to check duplicate users, we can iterate through the Map (takes constant time).